docs/architecture/ingest.md

```diff
@@ -0,0 +1,87 @@
+An example of a listener is the healtcheck component. Once it is notified of the price update, it
+- Pool Liquidity Pricing Worker: This worker is responsible for updating the pool liquidity capitalization
+based on the prices that are computed by the pricing worker.
+
+### Pool Liquidity Pricing
+
+The pool liquidity pricing worker is responsible for computing liquidity capitalization (configured in USDC by default).
+
+There are 2 kinds of liquidities capitalizations that we compute:
+1. Token liquidity capitalization across all pools
+  * Value of all pool liquidity of a token across all pools
+2. Pool liqudity capitalization
+  * Value of all pool balances in USDC
+
+To compute the above, we need to know:
+1. Prices for all tokens
+2. Liquidity for all tokens in all pools
+
+There are 2 possble cases:
+1. All pools are being ingested and processed
+   * This happens at cold start or if any error is returned in a previois block.
+2. Only the pools that have changed within a height are being ingested
+   * This assumes that all other pools have already been ingested in a previous height.
+
+
+In the current system, When we push only a subset of pools, we do not need to read all ~1700 other pools. However, to commpute denom liquidity capitalization across all pools, we might need to know all pools that are associated with a particular token.
+
+Therefore, we store the following data structure:
+
+```go
+// DenomPoolLiquidityData contains the pool liquidity data for a denom
+// It has the total liquidity for the denom as well as all the
+// pools with their individual contributions to the total.
+type DenomPoolLiquidityData struct {
+	// Total liquidity for this denom
+	TotalLiquidity osmomath.Int
+	// Mapping from pool ID to denom liquidity
+	// in that pool
+	Pools map[uint64]osmomath.Int
+}
+
+// DenomPoolLiquidityMap is a map of denoms to their pool liquidity data.
+type DenomPoolLiquidityMap map[string]DenomPoolLiquidityData
+```
+
+This allows us to identify all pools that are associated with a particular token and read their liquidity
+data without having to retrieve the entire pool model (or even read all pool liquidity values). Assume pool
+1 is updated within a block and has denom ATOM. Then, we get `DenomPoolLiquidityData` for ATOM, subtract
+the old liqudity contribution of that pool to `TotalLiquidity` and, finally add the new contributions to the total while also updating the entry in the `Pools` map.
+
+In such a way, we only had to do 2 map accesses and 3 math operations while updating the liquidity for a token.
+
+We [store this map in-memory of the ingester module](https://github.com/osmosis-labs/sqs/blob/81452a23b12fe9744e30ee04f5c13c790e404e51/ingest/usecase/ingest_usecase.go#L177), updating it while processing each block.
```

Comment:
When I read this I start to feel there is a bit of disconnected information here, maybe it may be better once I see the example of data being pushed from the node. The general flow I got is:

All pool data OR changed pool data pushed to SQS => we update DenomPoolLiquidityMap (well written in this doc) => and then how token prices are then computed ? Let's say the case is ATOM, we have pools of ATOM/OSMO, ATOM/USDC, ATOM/AKT, how do we compute the price after updating DenomPoolLiquidityMap?

Reply:
This is a pool liquidity pricing worker that computes pool liquidity capitalization.

For example, assume that there is an ATOM/OSMO pool that is modified within a block. First, we recompute the default quote denom (USDC) denominated prices for ATOM and OSMO using the "pricing worker". Then, the "pool liquidity pricing worker" uses the updated prices from the pricing worker to recompute the capitalization (USDC-denominated value of total liquidity in the pool).

Reply:
Added this example here:
[37d0db3](https://github.com/osmosis-labs/sqs/pull/281/commits/37d0db3a5b76fea4c7e4f55e7c5762788972ee01)

Please feel free to recommend edits to make this clear

Reply:
Thanks! That clarifies a lot to me, and sorry for asking simple questions like that! I'm trying to read this doc from a fresh eye's perspective. 

For "pricing worker" which computes the token price (USDC quote denom) - is it really "computed" by the SQS? or such data is part of the pushed data from the node? I remember I worked on a method before (in the technical challenge) called "pool.SpotPrice(ctx, denom1, denom2)" and that's why I wonder such data is from the node or computed in sqs.

Reply:
Explained in a conversation with Roman. Thanks!